controllers/instanceha/instanceha_controller.go

```diff
@@ -0,0 +1,576 @@
+		condition.UnknownCondition(instancehav1.InstanceHAReadyCondition, condition.InitReason, instancehav1.InstanceHAReadyInitMessage),
+		// service account, role, rolebinding conditions
+		condition.UnknownCondition(condition.ServiceAccountReadyCondition, condition.InitReason, condition.ServiceAccountReadyInitMessage),
+		condition.UnknownCondition(condition.RoleReadyCondition, condition.InitReason, condition.RoleReadyInitMessage),
+		condition.UnknownCondition(condition.RoleBindingReadyCondition, condition.InitReason, condition.RoleBindingReadyInitMessage),
+		condition.UnknownCondition(condition.NetworkAttachmentsReadyCondition, condition.InitReason, condition.NetworkAttachmentsReadyInitMessage),
+	)
+	instance.Status.Conditions.Init(&cl)
+	instance.Status.ObservedGeneration = instance.Generation
+
+	//// mark
+	if instance.Status.NetworkAttachments == nil {
+		instance.Status.NetworkAttachments = map[string][]string{}
+	}
+
+	// Service account, role, binding
+	rbacRules := []rbacv1.PolicyRule{
+		{
+			APIGroups:     []string{"security.openshift.io"},
+			ResourceNames: []string{"anyuid"},
+			Resources:     []string{"securitycontextconstraints"},
+			Verbs:         []string{"use"},
+		},
+		{
+			APIGroups: []string{""},
+			Resources: []string{"pods"},
+			Verbs:     []string{"create", "get", "list", "watch", "update", "patch", "delete"},
+		},
+	}
+	rbacResult, err := common_rbac.ReconcileRbac(ctx, helper, instance, rbacRules)
+	if err != nil {
+		return rbacResult, err
+	} else if (rbacResult != ctrl.Result{}) {
+		return rbacResult, nil
+	}
+
+	//
+	// Validate that keystoneAPI is up
+	//
+	keystoneAPI, err := keystonev1.GetKeystoneAPI(ctx, helper, instance.Namespace, map[string]string{})
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			instance.Status.Conditions.Set(condition.FalseCondition(
+				instancehav1.InstanceHAReadyCondition,
+				condition.RequestedReason,
+				condition.SeverityInfo,
+				instancehav1.InstanceHAKeystoneWaitingMessage))
+			Log.Info("KeystoneAPI not found!")
+			return ctrl.Result{RequeueAfter: time.Duration(5) * time.Second}, nil
+		}
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			instancehav1.InstanceHAReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			instancehav1.InstanceHAReadyErrorMessage,
+			err.Error()))
+		return ctrl.Result{}, err
+	}
+	if !keystoneAPI.IsReady() {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			instancehav1.InstanceHAReadyCondition,
+			condition.RequestedReason,
+			condition.SeverityInfo,
+			instancehav1.InstanceHAKeystoneWaitingMessage))
+		Log.Info("KeystoneAPI not yet ready")
+		return ctrl.Result{RequeueAfter: time.Duration(5) * time.Second}, nil
+	}
+
+	Labels := map[string]string{
+		common.AppSelector: "instanceha",
+	}
+
+	configVars := make(map[string]env.Setter)
+
+	_, configMapHash, err := configmap.GetConfigMapAndHashWithName(ctx, helper, *instance.Spec.OpenStackConfigMap, instance.Namespace)
```

Comment:
if we keep it to be a ptr in the api, should probably check if it is != nil, even it is a required value